plugins/inputs/mesos/mesos.go

```diff
@@ -593,6 +621,29 @@ func (m *Mesos) gatherMainMetrics(u *url.URL, role Role, acc telegraf.Accumulato
 		}
 	}
 
+	var includeFrameworkOffers, includeAllocator bool
+	for _, col := range m.MasterCols {
+		if col == "framework_offers" {
+			includeFrameworkOffers = true
+		} else if col == "allocator" {
+			includeAllocator = true
+		}
+	}
+
+	for metricName := range jf.Fields {
+		if !strings.HasPrefix(metricName, "master/frameworks/") && !strings.HasPrefix(metricName, "frameworks/") && !strings.HasPrefix(metricName, "allocator/") {
+			continue
+		}
+
+		// filter out framework offers/allocator metrics if necessary
+		if !includeFrameworkOffers &&
+			(strings.HasPrefix(metricName, "master/frameworks/") || strings.HasPrefix(metricName, "frameworks/")) ||
+			(!includeAllocator && strings.HasPrefix(metricName, "allocator/")) {
```

Comment:
This logic ended up a bit too complex for my taste, can you move the filtering into `func (m *Mesos) filterMetrics(`?  I think we could wedge in a switch, something like this:
```
	for _, k := range metricsDiff(role, selectedMetrics) {
		fmt.Println("group:", k)
		switch k {
		case "allocators":
			// TODO remove all allocator metrics
		case "framework_offers":
			// TODO remove all framework_offers
		default:
            // Removes other categories
			for _, v := range getMetrics(role, k) {
				if _, ok = (*metrics)[v]; ok {
					delete((*metrics), v)
				}
			}
		}
	}
```

Reply:
Pushed 658414e. This required a refactor to the test module: I had to move `masterMetricNames` out of `generateMetrics()` and make it a global, so that `TestMasterFilter()` could know which metrics to expect. I did the same with `slaveMetricNames` just for consistency. LMK what you think!